src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,473 @@
+
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), MaxSubscriptionSize());
+
+    ReturnErrorOnFailure(Save(writer, subscriptionInfo));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+
+    ReturnErrorOnFailure(
+        mStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumption(firstEmptySubscriptionIndex).KeyName(),
+                                  backingBuffer.Get(), static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::Delete(NodeId nodeId, FabricIndex fabricIndex, SubscriptionId subscriptionId)
+{
+    bool subscriptionFound   = false;
+    CHIP_ERROR lastDeleteErr = CHIP_NO_ERROR;
+
+    uint16_t remainingSubscriptionsCount = 0;
+    for (uint16_t subscriptionIndex = 0; subscriptionIndex < CHIP_IM_MAX_NUM_SUBSCRIPTIONS; subscriptionIndex++)
+    {
+        SubscriptionInfo subscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, subscriptionInfo);
+
+        // delete match
+        if (err == CHIP_NO_ERROR)
+        {
+            if ((nodeId == subscriptionInfo.mNodeId) && (fabricIndex == subscriptionInfo.mFabricIndex) &&
+                (subscriptionId == subscriptionInfo.mSubscriptionId))
+            {
+                subscriptionFound    = true;
+                CHIP_ERROR deleteErr = Delete(subscriptionIndex);
+                if (deleteErr != CHIP_NO_ERROR)
+                {
+                    lastDeleteErr = deleteErr;
+                }
+            }
+            else
+            {
+                remainingSubscriptionsCount++;
+            }
+        }
+    }
+
+    // if there are no persisted subscriptions, the MaxCount can also be deleted
+    if (remainingSubscriptionsCount == 0)
+    {
+        DeleteMaxCount();
+    }
+
+    if (lastDeleteErr != CHIP_NO_ERROR)
+    {
+        return lastDeleteErr;
+    }
+
+    return subscriptionFound ? CHIP_NO_ERROR : CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::DeleteMaxCount()
+{
+    return mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionMaxCount().KeyName());
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::DeleteAll(FabricIndex fabricIndex)
+{
+    CHIP_ERROR deleteErr = CHIP_NO_ERROR;
+
+    uint16_t count = 0;
+    for (uint16_t subscriptionIndex = 0; subscriptionIndex < CHIP_IM_MAX_NUM_SUBSCRIPTIONS; subscriptionIndex++)
+    {
+        SubscriptionInfo subscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, subscriptionInfo);
+
+        if (err == CHIP_NO_ERROR)
+        {
+            if (fabricIndex == subscriptionInfo.mFabricIndex)
+            {
+                err = Delete(subscriptionIndex);
+                if ((err != CHIP_NO_ERROR) && (err != CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND))
```

Comment:
How could we have CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND here, if we just loaded that index above?